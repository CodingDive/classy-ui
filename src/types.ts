export interface IClass {
  id: string;
  classname: string;
  variant: string | null;
  shortName: string;
  css: (name: string) => string;
  variable: {
    value: string;
    originalValue: string;
  } | null;
}

export type IClasses = Record<string, IClass>;

export type IVariants = Record<string, string>;

export type IClassnames<T extends string> = Record<
  string,
  {
    variants?:
      | { [name: string]: string }
      | ((
          variables: IVariables<T>,
          utils: {
            negative: (value: { [key: string]: string }) => { [key: string]: string };
          },
        ) => IVariants);
    css: (name: string, value: string) => string;
  }
>;

export interface IEvaluatedClassnames {
  [name: string]: {
    variants: { [name: string]: string };
    variantsWithoutVariables: { [name: string]: string };
    css: (name: string, value: string) => string;
  };
}

export type IVariables<T extends string> = {
  [key in T]: { [variant: string]: string };
};

export interface IBaseConfig<T extends string, U = IVariables<T>> {
  variables: IVariables<T>;
  screens: {
    [name: string]: (css: string, variables: IVariables<T>) => string;
  };
  classnames: IClassnames<T>;
  themes?: {
    [name: string]: {
      [key in keyof U]: U[key];
    };
  };
}

export interface IConfig<T extends string, U = IVariables<T>> {
  variables?: {
    [key in T]: ((variables: IVariables<T>) => { [variant: string]: string }) | { [variant: string]: string };
  };
  screens?: {
    [name: string]: (css: string, variables: IVariables<T>) => string;
  };
  classnames?: IClassnames<T>;
  themes?: {
    [name: string]: {
      [key in keyof U]: U[key];
    };
  };
}

export interface IEvaluatedThemes {
  [variable: string]: {
    [key: string]: {
      [theme: string]: string;
    };
  };
}

export interface IEvaluatedConfig {
  variables: IVariables<any>;
  screens: {
    [name: string]: (css: string, variables: IVariables<any>) => string;
  };
  classnames: IEvaluatedClassnames;
  themeNames: string[];
  themes?: IEvaluatedThemes;
}

export interface IClassesByType {
  screens: {
    [type: string]: string[];
  };
  common: {
    [id: string]: string;
  };
  themeVariables: {
    [theme: string]: {
      [variable: string]: string;
    };
  };
  rootVariables: {
    [variable: string]: string;
  };
}

export interface IExtractedClass {
  id?: string;
  uid?: string;
  name: string;
  decorators: string[];
}

export interface IExtractedClasses {
  [uid: string]: IExtractedClass;
}
