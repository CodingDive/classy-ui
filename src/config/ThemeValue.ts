import { IConfigDefaults } from '../types';

export class ThemeValue {
  category: keyof IConfigDefaults;
  label: string;
  themes: string[];
  value: string;
  constructor(category: keyof IConfigDefaults, label: string, value: string, themes: string[]) {
    this.category = category;
    this.label = label;
    this.value = value;
    this.themes = themes;
  }
  copy(value: string) {
    return new ThemeValue(this.category, this.label, value, this.themes);
  }
}
