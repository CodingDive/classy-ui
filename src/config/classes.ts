import { IConfig, TClasses } from '../types';

export const classes = (
  config: IConfig,
): {
  [key in TClasses]?: string;
} => ({
  'bg-transparent': `{background-color: ${config.backgrounds.backgroundColor.transparent}`,
  'bg-black': `{background-color: ${config.backgrounds.backgroundColor.black}`,
  'bg-white': `{background-color: ${config.backgrounds.backgroundColor.white}`,
  'bg-gray-100': `{background-color: ${config.backgrounds.backgroundColor['gray-100']}`,
  'bg-gray-200': `{background-color: ${config.backgrounds.backgroundColor['gray-200']}`,
  'bg-gray-300': `{background-color: ${config.backgrounds.backgroundColor['gray-300']}`,
  'bg-gray-400': `{background-color: ${config.backgrounds.backgroundColor['gray-400']}`,
  'bg-gray-500': `{background-color: ${config.backgrounds.backgroundColor['gray-500']}`,
  'bg-gray-600': `{background-color: ${config.backgrounds.backgroundColor['gray-600']}`,
  'bg-gray-700': `{background-color: ${config.backgrounds.backgroundColor['gray-700']}`,
  'bg-gray-800': `{background-color: ${config.backgrounds.backgroundColor['gray-800']}`,
  'bg-gray-900': `{background-color: ${config.backgrounds.backgroundColor['gray-900']}`,
});
