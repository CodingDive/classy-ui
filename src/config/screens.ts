import { IScreens } from '../types';

const screens: IScreens = {
  mobile: (css, breakpoints) => `@media (min-width: ${breakpoints.MOBILE}) {${css}}`,
  tablet: (css, breakpoints) => `@media (min-width: ${breakpoints.TABLET}) {${css}}`,
  laptop: (css, breakpoints) => `@media (min-width: ${breakpoints.LAPTOP}) {${css}}`,
  desktop: (css, breakpoints) => `@media (min-width: ${breakpoints.DESKTOP}) {${css}}`,
};

export default screens;
