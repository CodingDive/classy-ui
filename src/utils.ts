import { join } from 'path';

// @ts-ignore
import reduceCalc from 'reduce-css-calc';

import { config as baseConfig } from './config/base.config';
import {
  IClasses,
  IClassesByType,
  IClassnames,
  IConfig,
  IEvaluatedClassnames,
  IEvaluatedConfig,
  IEvaluatedThemes,
  IExtractedClass,
  IExtractedClasses,
  IGlobalTokens,
} from './types';

export const allowedPseudoDecorators = [
  'hover',
  'focus',
  'active',
  'disabled',
  'visited',
  'firstChild',
  'lastChild',
  'oddChild',
  'evenChild',
  'focusWithin',
];

export const getClassesFromConfig = (
  classnameKey: string,
  config: IEvaluatedConfig,
  getShortName: (labelIndex: number) => string,
) => {
  const classname = config.classnames[classnameKey];

  return Object.keys(classname.tokens).reduce((aggr, tokenKey, tokenIndex) => {
    const id = `${camelToDash(classnameKey)}-${tokenKey}`;
    const cssDeriver = config.classnames[classnameKey].css;
    return {
      ...aggr,
      [id]: {
        id,
        classname: classnameKey,
        token: tokenKey,
        derived: Array.isArray(cssDeriver) ? cssDeriver : null,
        variable:
          classname.tokens[tokenKey] !== classname.tokensWithoutVariables[tokenKey]
            ? {
                value: classname.tokens[tokenKey],
                originalValue: classname.tokensWithoutVariables[tokenKey],
              }
            : null,
        shortName: getShortName(tokenIndex),
      },
    };
  }, {} as IClasses);
};

export const deepAssign = (
  a: { [key: string]: { [key: string]: string } },
  b: { [key: string]: { [key: string]: string } },
) => {
  Object.keys(b).forEach(key => {
    if (!a[key]) {
      a[key] = {};
    }
    Object.keys(b[key]).forEach(subKey => {
      a[key][subKey] = b[key][subKey];
    });
  });

  return a;
};

export const evaluateConfig = (config: IConfig<any>): IEvaluatedConfig => {
  const configTokens = config.tokens || {};
  const originalTokens: IGlobalTokens<any> = {
    ...baseConfig.tokens,
    ...(Object.keys(configTokens).reduce<IGlobalTokens<any>>((aggr, key) => {
      if (typeof configTokens[key] === 'function') {
        aggr[key] = (configTokens[key] as any)(baseConfig.tokens);
      } else {
        aggr[key] = configTokens[key] as any;
      }
      return aggr;
    }, {}) as any),
  };

  // Reverse themes lookup to tokens instead
  const configThemes = config.themes || {};
  const themesByTokens = Object.keys(configThemes).reduce((aggr, themeKey) => {
    Object.keys(configThemes[themeKey]).forEach(tokenKey => {
      if (!aggr[tokenKey]) {
        aggr[tokenKey] = {};
      }
      Object.keys(configThemes[themeKey][tokenKey]).forEach(valueKey => {
        if (!aggr[tokenKey][valueKey]) {
          aggr[tokenKey][valueKey] = {};
        }
        aggr[tokenKey][valueKey][themeKey] = configThemes[themeKey][tokenKey][valueKey];
      });
    });

    return aggr;
  }, {} as IEvaluatedThemes);

  // Evaluated variables where values are replaced by CSS variable
  const tokens = Object.keys(originalTokens).reduce<IGlobalTokens<any>>((aggr, key) => {
    aggr[key] = Object.keys(originalTokens[key]).reduce<{ [token: string]: string }>((subAggr, subKey) => {
      subAggr[subKey] =
        themesByTokens[key] && themesByTokens[key][subKey] ? `var(--${key}-${subKey})` : originalTokens[key][subKey];

      return subAggr;
    }, {});

    return aggr;
  }, {});

  // Call any dynamic classname tokens with both the original variables and
  // the ones who have been evaluated with CSS variables
  const allClassnames: IClassnames<any> = { ...baseConfig.classnames, ...config.classnames };
  const classnames = Object.keys(allClassnames).reduce((aggr, key) => {
    if (typeof allClassnames[key] === 'function') {
      aggr[key] = allClassnames[key] as any;
    } else {
      aggr[key] = {
        ...allClassnames[key],
        tokensWithoutVariables:
          typeof (allClassnames[key] as any).tokens === 'function'
            ? (allClassnames[key] as any).tokens(originalTokens, { negative })
            : (allClassnames[key] as any).tokens,
        tokens:
          typeof (allClassnames[key] as any).tokens === 'function'
            ? (allClassnames[key] as any).tokens(tokens, { negative })
            : (allClassnames[key] as any).tokens,
        description: (allClassnames[key] as any).description,
      } as any;
    }

    return aggr;
  }, {} as IEvaluatedClassnames);

  return {
    tokens,
    screens: config.screens || baseConfig.screens,
    classnames,
    themes: themesByTokens,
    themeNames: Object.keys(config.themes || {}),
  };
};

export const getUserConfig = () => {
  try {
    return require(join(process.cwd(), 'classy-ui.config.js'));
  } catch (error) {
    return {};
  }
};

export const createProductionCss = (productionClassesByType: IClassesByType, config: IEvaluatedConfig) => {
  let css = Object.keys(productionClassesByType.common).reduce(
    (aggr, name) => aggr + productionClassesByType.common[name],
    '',
  );

  const screenKeys = Object.keys(productionClassesByType.screens);
  screenKeys.forEach(screen => {
    if (productionClassesByType.screens[screen].length) {
      const screenCss = productionClassesByType.screens[screen].reduce((aggr, classCss) => {
        return aggr + classCss;
      }, '');
      css += config.screens[screen](screenCss, config.tokens);
    }
  });

  const variableKeys = Object.keys(productionClassesByType.rootTokens);

  if (variableKeys.length) {
    css += ':root{';
    variableKeys.forEach(key => {
      css += `${key}:${productionClassesByType.rootTokens[key]};`;
    });
    css += '}';
  }

  Object.keys(productionClassesByType.themeTokens).forEach(theme => {
    const variables = Object.keys(productionClassesByType.themeTokens[theme]).reduce(
      (aggr, variableKey) => `${aggr}${productionClassesByType.themeTokens[theme][variableKey]}`,
      '',
    );
    css += `.themes-${theme}{${variables}}`;
  });

  return css;
};

export const camelToDash = (string: string) => {
  return string
    .replace(/[\w]([A-Z])/g, function(m) {
      return m[0] + '-' + m[1];
    })
    .toLowerCase();
};

export const createClassEntry = (name: string, decorators: string[], css: (name: string) => string) => {
  const groupDecorators = decorators
    .filter(decorator => decorator.startsWith('group') && decorator !== 'group')
    .map(decorator => camelToDash(decorator.substr(5)));
  const pseudoDecorators = decorators
    .filter(decorator => allowedPseudoDecorators.includes(decorator))
    .map(decorator => camelToDash(decorator));
  const evaluatedName = `.${name.replace(/\:/g, '\\:')}${
    pseudoDecorators.length ? `:${pseudoDecorators.join(':')}` : ''
  }`;

  return `${groupDecorators.length ? `.group:${groupDecorators.join(':')} ` : ''}${css(evaluatedName)}`;
};

export const flat = (array: any[]) => array.reduce((aggr, item) => aggr.concat(item), []);

export const injectProduction = (classCollection: IExtractedClasses, classes: IClasses, config: IEvaluatedConfig) => {
  const productionClassesByType: IClassesByType = {
    screens: {},
    common: {},
    themeTokens: {},
    rootTokens: {},
  };

  Object.keys(classCollection).forEach(uid => {
    const extractedClass = classCollection[uid];
    const configClass = classes[extractedClass.id as string];
    const screenDecorators = extractedClass.decorators.filter(decorator => decorator in config.screens);
    const otherDecorators = extractedClass.decorators.filter(decorator => !(decorator in config.screens));
    let classEntry: any;
    try {
      const prefix = extractedClass.name.substr(0, extractedClass.name.lastIndexOf(':') + 1);
      const cssProcessor = config.classnames[configClass.classname].css;
      // The classname definition might references other CSS processors, we extract the base classname for lookups
      const classnameKeys = Array.isArray(cssProcessor) ? cssProcessor : [configClass.classname];

      classnameKeys.forEach(classnameKey => {
        const classConfig = config.classnames[classnameKey];
        const id = `${camelToDash(classnameKey)}-${configClass.token}`;
        const name = classes[id].shortName;
        const classname = prefix + name.replace('@', '\\@');

        classEntry = createClassEntry(classname, otherDecorators, evaluatedName =>
          (classConfig.css as any)(evaluatedName, classConfig.tokens[configClass.token]),
        );

        if (screenDecorators.length) {
          screenDecorators.forEach(screen => {
            productionClassesByType.screens[screen] = productionClassesByType.screens[screen] || [];
            productionClassesByType.screens[screen].push(classEntry);
          });
        } else {
          productionClassesByType.common[classname] = classEntry;
        }

        if (configClass.variable) {
          const themes = config.themes || {};
          const variableValue = configClass.variable.value;
          const originalValue = configClass.variable.originalValue;
          const variables = (variableValue.match(/var\(.*\)/) || []).map(varString =>
            varString.replace(/var\(|\)/g, ''),
          );

          config.themeNames.forEach(theme => {
            productionClassesByType.themeTokens[theme] = productionClassesByType.themeTokens[theme] || {};
            variables.forEach(variable => {
              const variableParts = variable.substr(2).split('-');
              const variableKey = variableParts.shift() as string;
              const variableValueKey = variableParts.join('-');

              productionClassesByType.themeTokens[theme][
                variable
              ] = `${variable}:${themes[variableKey][variableValueKey][theme]};`;
              productionClassesByType.rootTokens[variable] = originalValue;
            });
          });
        }
      });
    } catch (error) {
      throw new Error(uid + JSON.stringify(extractedClass, null, 2));
    }
  });

  return createProductionCss(productionClassesByType, config);
};

export const injectDevelopment = (classCollection: IExtractedClasses, classes: IClasses, config: IEvaluatedConfig) => {
  return Object.keys(classCollection).reduce((aggr, uid) => {
    const extractedClass = classCollection[uid];
    const mainId = extractedClass.id as string;
    const screenDecorators = extractedClass.decorators.filter(decorator => decorator in config.screens);
    const otherDecorators = extractedClass.decorators.filter(decorator => !(decorator in config.screens));
    const prefix = extractedClass.name.substr(0, extractedClass.name.lastIndexOf(':') + 1);
    const configClass = classes[mainId];
    const cssProcessor = config.classnames[configClass.classname].css;
    // The classname definition might references other CSS processors, we extract the base classname for lookups
    const classnameKeys = Array.isArray(cssProcessor) ? cssProcessor : [configClass.classname];

    return aggr.concat(
      classnameKeys.reduce((injections, classnameKey) => {
        const classConfig = config.classnames[classnameKey];
        const name = `${camelToDash(classnameKey)}@${configClass.token}`;
        const classname = prefix + name;
        const classEntry = createClassEntry(classname.replace('@', '\\@'), otherDecorators, evaluatedName =>
          (classConfig.css as any)(evaluatedName, classConfig.tokens[configClass.token]),
        );

        let css = '';

        if (screenDecorators.length) {
          screenDecorators.forEach(screen => {
            css += config.screens[screen](classEntry, config.tokens);
          });
        } else {
          css = classEntry;
        }

        if (configClass.variable) {
          const themes = config.themes || {};
          const variableValue = configClass.variable.value;
          const originalValue = configClass.variable.originalValue;
          const variables = (variableValue.match(/var\(.*\)/) || []).map(varString =>
            varString.replace(/var\(|\)/g, ''),
          );

          variables.forEach(variable => {
            const variableParts = variable.substr(2).split('-');
            const variableKey = variableParts.shift() as string;
            const variableValueKey = variableParts.join('-');

            config.themeNames.forEach(theme => {
              css += `:root{${variable}:${originalValue};}\n.themes-${theme}{${variable}:${themes[variableKey][variableValueKey][theme]};}`;
            });
          });
        }

        return injections.concat([classname, css]);
      }, [] as string[]),
    );
  }, [] as string[]);
};

export const negative = (scale: { [key: string]: string }) => {
  return Object.keys(scale)
    .filter(key => scale[key] !== '0')
    .reduce(
      (negativeScale, key) => ({
        ...negativeScale,
        [`NEG_${key}`]: negateValue(scale[key]),
      }),
      {},
    );
};

export const negateValue = (value: string) => {
  try {
    return reduceCalc(`calc(${value} * -1)`);
  } catch (e) {
    return value;
  }
};

export const createClassObject = (
  {
    baseClass,
    token,
    decorators,
  }: {
    baseClass: string;
    token: string;
    decorators: string[];
  },
  classes: IClasses,
  isProduction: boolean,
): IExtractedClass => {
  const id = `${camelToDash(baseClass)}-${token}`;
  const uid = [decorators.sort().join(':'), `${camelToDash(baseClass)}@${token}`]
    .filter(Boolean)
    .filter(i => i!.length > 0)
    .join(':');

  const returnedDecorators = decorators.slice() as IExtractedClass['decorators'];

  let name: string;
  if (id && isProduction && classes[id].derived) {
    name = classes[id]
      .derived!.reduce((aggr, key) => {
        return aggr.concat(classes[`${camelToDash(key)}-${token}`].shortName);
      }, [] as string[])
      .join(' ');
  } else if (id && isProduction) {
    name = classes[id].shortName;
  } else if (id && classes[id].derived) {
    name = classes[id]
      .derived!.reduce((aggr, key) => {
        return aggr.concat(`${camelToDash(key)}@${token}`);
      }, [] as string[])
      .join(' ');
  } else {
    name = uid;
  }

  return {
    id,
    uid,
    name,
    decorators: returnedDecorators,
  };
};

export const generateShortName = (num: number) => {
  const baseChar = 'A'.charCodeAt(0);
  let letters = '';

  do {
    num -= 1;
    letters = String.fromCharCode(baseChar + (num % 26)) + letters;
    num = (num / 26) >> 0;
  } while (num > 0);

  return letters;
};

export const hyphenToCamelCase = (str: string) =>
  str.replace(/-([a-z])/g, function(g) {
    return g[1].toUpperCase();
  });
