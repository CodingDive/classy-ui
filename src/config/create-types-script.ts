import { writeFileSync } from 'fs';
import { join } from 'path';

import { evaluateConfig } from '../utils';

import { transform as transformClassesToTypes } from './transform-classes-to-types';

const config = evaluateConfig({});

if (process.env.NODE_ENV !== 'test') {
  const libTypesPath = join(process.cwd(), 'lib', 'classy-ui.d.ts');
  const types = transformClassesToTypes(config);
  writeFileSync(libTypesPath, types);
}
