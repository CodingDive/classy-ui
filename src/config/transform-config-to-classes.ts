import { IClasses, IConfig, IEvaluatedConfig, TCssClasses } from '../types';
import { generateShortName, getClassesFromConfig } from '../utils';

const cssClasses: TCssClasses = {
  padding: ['padding', 'padding-top', 'padding-right', 'padding-bottom', 'padding-left'],
  margin: ['margin', 'margin-top', 'margin-right', 'margin-bottom', 'margin-left'],
  placeholderColor: [],
  gap: ['gap'],
  rowGap: ['row-gap'],
  columnGap: ['column-gap'],
  gridTemplateColumns: ['grid-template-columns'],
  gridColumnStart: ['grid-column-start'],
  gridColumnEnd: ['grid-column-end'],
  gridTemplateRows: ['grid-template-rows'],
  gridRow: ['grid-row'],
  gridRowStart: ['grid-row-start'],
  gridRowEnd: ['grid-row-end'],
  transformOrigin: ['transform-origin'],
  scale: ['scale'],
  rotate: ['rotate'],
  translate: ['translate'],
  skew: ['skew'],
  transitionProperty: ['transition-property'],
  transitionTimingFunction: ['transition-timing-function'],
  transitionDuration: ['transition-duration'],
  gridColumn: ['grid-column'],
  width: ['width'],
  borderWidth: ['border-width', 'border-top-width', 'border-right-width', 'border-bottom-width', 'border-left-width'],
  borderColor: ['border-color', 'border-bottom-color', 'border-top-color', 'border-left-color', 'border-right-color'],
  borderRadius: [
    'border-radius',
    'border-top-left-radius',
    'border-top-right-radius',
    'border-bottom-right-radius',
    'border-bottom-left-radius',
  ],
  fontSize: ['font-size'],
  fontWeight: ['font-weight'],
  fontFamily: ['font-family'],
  letterSpacing: ['letter-spacing'],
  lineHeight: ['line-height'],
  boxShadow: ['box-shadow', 'text-shadow'],
  zIndex: ['z-index'],
  backgroundPosition: ['background-position'],
  backgroundSize: ['background-size'],
  cursor: ['cursor'],
  fill: ['fill'],
  flex: ['flex'],
  inset: ['inset'],
  listStyleType: ['list-style-type'],
  objectPosition: ['object-position'],
  opacity: ['opacity'],
  order: ['order'],
  stroke: ['stroke'],
  strokeWidth: ['stroke-width'],
  backgroundColor: ['background-color'],
  container: [],
  flexGrow: ['flex-grow'],
  flexShrink: ['flex-shrink'],
  height: ['height'],
  maxHeight: ['max-height'],
  maxWidth: ['max-width'],
  minHeight: ['min-height'],
  minWidth: ['min-width'],
  color: ['color'],
};

export const transform = (config: IEvaluatedConfig): { defaults: IClasses; themes: { [name: string]: boolean } } => {
  const keys = Object.keys(cssClasses) as Array<keyof TCssClasses>;
  let cssPropertyCount = 0;
  const defaults = keys.reduce((aggr, key) => {
    return {
      ...aggr,
      ...getClassesFromConfig(
        key,
        config,
        cssClasses[key],
        (labelIndex: number) => `${generateShortName(cssPropertyCount++)}-${generateShortName(labelIndex)}`,
      ),
    };
  }, {} as IClasses);

  return {
    defaults,
    themes: Object.keys(config.themes || {}).reduce((aggr, theme) => {
      aggr[`themes-${theme}`] = true;

      return aggr;
    }, {} as any),
  };
};
