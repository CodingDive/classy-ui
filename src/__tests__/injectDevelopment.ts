import { config as baseConfig } from '../config/base.config';
import { transform } from '../config/transform-config-to-classes';
import { IExtractedClasses } from '../types';
import { injectDevelopment, mergeConfigs } from '../utils';

const config = mergeConfigs(baseConfig, {
  themes: {
    dark: {
      colors: {
        'red-700': 'green',
      },
    },
  },
});
const classes = transform(config);

function createExtractedClass(id: string, decorators: string[] = []) {
  return {
    id,
    uid: id,
    name: id,
    decorators,
  };
}

describe('INJECT DEVELOPMENT', () => {
  test('should inject simple', () => {
    const classCollection: IExtractedClasses = {
      ['background-color-red-500']: createExtractedClass('background-color-red-500'),
    };

    expect(injectDevelopment(classCollection, classes, config)).toMatchSnapshot();
  });

  test('should inject with pseudo selector', () => {
    const classCollection: IExtractedClasses = {
      ['background-color-red-500']: createExtractedClass('background-color-red-500', ['classnames', 'hover']),
    };

    expect(injectDevelopment(classCollection, classes, config)).toMatchSnapshot();
  });
  test('should inject with breakpoint', () => {
    const classCollection: IExtractedClasses = {
      ['background-color-red-500']: createExtractedClass('background-color-red-500', ['classnames', 'md']),
    };

    expect(injectDevelopment(classCollection, classes, config)).toMatchSnapshot();
  });
  test('should inject both pseudo selector and breakpoint', () => {
    const classCollection: IExtractedClasses = {
      ['background-color-red-500']: createExtractedClass('background-color-red-500', ['classnames', 'md', 'hover']),
    };

    expect(injectDevelopment(classCollection, classes, config)).toMatchSnapshot();
  });
  test('should inject with multiple pseudo selectors', () => {
    const classCollection: IExtractedClasses = {
      ['background-color-red-500']: createExtractedClass('background-color-red-500', [
        'classnames',
        'hover',
        'firstChild',
      ]),
    };

    expect(injectDevelopment(classCollection, classes, config)).toMatchSnapshot();
  });
  test('should inject with multiple breakpoints', () => {
    const classCollection: IExtractedClasses = {
      ['background-color-red-500']: createExtractedClass('background-color-red-500', ['classnames', 'sm', 'xl']),
    };

    expect(injectDevelopment(classCollection, classes, config)).toMatchSnapshot();
  });
  test('should inject multiple', () => {
    const classCollection: IExtractedClasses = {
      ['background-color-red-500']: createExtractedClass('background-color-red-500', ['classnames', 'md']),
      ['background-color-red-600']: createExtractedClass('background-color-red-600', ['classnames', 'hover', 'md']),
    };

    expect(injectDevelopment(classCollection, classes, config)).toMatchSnapshot();
  });
  test('should inject themes', () => {
    const classCollection: IExtractedClasses = {
      ['background-color-red-700']: createExtractedClass('background-color-red-700', ['classnames', 'md']),
    };

    expect(injectDevelopment(classCollection, classes, config)).toMatchSnapshot();
  });
  test('should inject group decorators', () => {
    const classCollection: IExtractedClasses = {
      ['background-color-red-700']: createExtractedClass('background-color-red-700', ['classnames', 'groupHover']),
    };

    expect(injectDevelopment(classCollection, classes, config)).toMatchSnapshot();
  });
});
