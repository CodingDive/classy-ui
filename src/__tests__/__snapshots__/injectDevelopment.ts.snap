// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`INJECT DEVELOPMENT should inject both pseudo selector and breakpoint 1`] = `
Array [
  "hover:md:color-red",
  "@media (max-width:768px) {.hover\\\\:md\\\\:color-red:hover{color:red;}}",
]
`;

exports[`INJECT DEVELOPMENT should inject group decorators 1`] = `
Array [
  "groupHover:color-green",
  ".group:hover .groupHover\\\\:color-green{color:var(--colors-green);}:root{--colors-green:green;}
.themes-dark{--colors-green:yellow;}",
]
`;

exports[`INJECT DEVELOPMENT should inject multiple 1`] = `
Array [
  "md:color-red",
  "@media (max-width:768px) {.md\\\\:color-red{color:red;}}",
  "hover:md:color-blue",
  "@media (max-width:768px) {.hover\\\\:md\\\\:color-blue:hover{color:blue;}}",
]
`;

exports[`INJECT DEVELOPMENT should inject simple 1`] = `
Array [
  "color-red",
  ".color-red{color:red;}",
]
`;

exports[`INJECT DEVELOPMENT should inject themes 1`] = `
Array [
  "md:color-green",
  "@media (max-width:768px) {.md\\\\:color-green{color:var(--colors-green);}}:root{--colors-green:green;}
.themes-dark{--colors-green:yellow;}",
]
`;

exports[`INJECT DEVELOPMENT should inject with breakpoint 1`] = `
Array [
  "md:color-red",
  "@media (max-width:768px) {.md\\\\:color-red{color:red;}}",
]
`;

exports[`INJECT DEVELOPMENT should inject with multiple breakpoints 1`] = `
Array [
  "sm:xl:color-red",
  "@media (max-width:640px) {.sm\\\\:xl\\\\:color-red{color:red;}}@media (max-width:1280px) {.sm\\\\:xl\\\\:color-red{color:red;}}",
]
`;

exports[`INJECT DEVELOPMENT should inject with multiple pseudo selectors 1`] = `
Array [
  "firstChild:hover:color-red",
  ".firstChild\\\\:hover\\\\:color-red:first-child:hover{color:red;}",
]
`;

exports[`INJECT DEVELOPMENT should inject with pseudo selector 1`] = `
Array [
  "hover:color-red",
  ".hover\\\\:color-red:hover{color:red;}",
]
`;
